File: Algorithms/yourname.github/HeapSort.cpp
#include<bits/stdc++.h>
using namespace std;
void heapify(int arr[], int n, int i)
{
    int largest = i; 
    int l = 2 * i + 1; 
    int r = 2 * i + 2; 

    if (l < n && arr[l] > arr[largest])
        largest = l;
 
    if (r < n && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        swap(arr[i], arr[largest]);
 
        heapify(arr, n, largest);
    }
}
 
void heapSort(int arr[], int n)
{

    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);
 
    for (int i = n - 1; i > 0; i--) {

        swap(arr[0], arr[i]);
 
        heapify(arr, i, 0);
    }
}
 
int main()
{
    int arr[5]={8,10,6,9,4};
    int n=5;
    cout<<"Before Sorting : "<<endl;
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    heapSort(arr,n);
    cout<<"After Sorting : "<<endl;
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    
    return 0;
}

File: Algorithms/yourname.github/HeapSort.py

"""
                                   Heap Sort
It is one of the efficient sorting algorithm based on heap data structure in O(nlogn).
Here the given array to be sorted is assumed to be a heap.So for ith index
The left child will become the element present at the 2*i+1 index in the array.
The right child will become the element present at the 2*i+2  index in the array.
Parent of the ith index will be element present at (i-1)/2 index in the array.
There are 2 major operations which are responsible for maintaining the heap Heapify:
->This operation sets the order of values between the parent and its child
->If we are dealing with the max heap, it will find the index having max value among the node and its children
->If the index holding max value is not the parent, it will wap the parent with the child having the max value
"""

def heapify(arr, n, i):
    largest = i  # Initialize largest as root
    l = 2 * i + 1     # left = 2*i + 1
    r = 2 * i + 2     # right = 2*i + 2

    # See if left child of root exists and is
    # greater than root
    if l < n and arr[i] < arr[l]:
        largest = l

    # See if right child of root exists and is
    # greater than root
    if r < n and arr[largest] < arr[r]:
        largest = r

    # Change root, if needed
    if largest != i:
        arr[i],arr[largest] = arr[largest],arr[i]  # 


        heapify(arr, n, largest)

def heapSort(arr):
    n = len(arr)


    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)


    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]   
        heapify(arr, i, 0)

# Driver code to test above

if __name__ == "__main__":
    arr = [eval(x) for x in input("Enter the array: ").split()]
    heapSort(arr)
    print(arr)